third_party/libusb/webport/src/libusb-to-webusb-adaptor.js

```diff
@@ -292,6 +303,160 @@ function getLibusbJsEndpointType(webusbEndpointType) {
+ * corresponding 'extraData' fields inside `libusbJsConfiguration`.
+ * @param {!Object} webusbDevice The WebUSB USBDevice value.
+ * @param {!LibusbJsConfigurationDescriptor} libusbJsConfiguration
+ */
+async function fetchAndFillConfigurationExtraData(
+    webusbDevice, libusbJsConfiguration) {
+  await webusbDevice['open']();
+  /** @preserveTry */
+  try {
+    await fetchAndFillConfigurationExtraDataForOpenedDevice(
+        webusbDevice, libusbJsConfiguration);
+  } finally {
+    // TODO(#429): Don't close if there are opened device handles.
+    await webusbDevice['close']();
+  }
+}
+
+/**
+ * Same as `fetchAndFillConfigurationExtraData()`, but assumes the device to be
+ * opened.
+ * @param {!Object} webusbDevice The WebUSB USBDevice value.
+ * @param {!LibusbJsConfigurationDescriptor} libusbJsConfiguration
+ */
+async function fetchAndFillConfigurationExtraDataForOpenedDevice(
+    webusbDevice, libusbJsConfiguration) {
+  const GET_DESCRIPTOR_REQUEST = 0x06;
+  const CONFIGURATION_DESCRIPTOR_TYPE = 0x02;
+  const INTERFACE_DESCRIPTOR_TYPE = 0x04;
+  const ENDPOINT_DESCRIPTOR_TYPE = 0x05;
+  const CONFIGURATION_DESCRIPTOR_LENGTH = 9;
+  const INTERFACE_DESCRIPTOR_LENGTH = 9;
+  const ENDPOINT_DESCRIPTOR_LENGTH = 7;
+
+  const controlRequestValue = (CONFIGURATION_DESCRIPTOR_TYPE << 8) |
+      (libusbJsConfiguration['configurationValue'] - 1);
+  const controlTransferParameters = {
+    'requestType': 'standard',
+    'recipient': 'device',
+    'request': GET_DESCRIPTOR_REQUEST,
+    'value': controlRequestValue,
+    'index': 0
+  };
+
+  // Determine the size of the whole descriptor hierarchy ("wTotalLength").
+  let transferResult = await webusbDevice['controlTransferIn'](
+      controlTransferParameters, CONFIGURATION_DESCRIPTOR_LENGTH);
+  if (transferResult['status'] !== 'ok' || !transferResult['data'])
+    return;
+  const initialData = /** @type {!DataView} */ (transferResult['data']);
+  const totalLength = initialData.getUint16(2, /*littleEndian=*/ true);
+
+  // Read the whole descriptor hierarchy.
+  transferResult = await webusbDevice['controlTransferIn'](
+      controlTransferParameters, totalLength);
+  if (transferResult['status'] !== 'ok' || !transferResult['data'])
+    return;
+  const descriptors = /** @type {!DataView} */ (transferResult['data']);
+  if (descriptors.byteLength !== totalLength)
+    return;
+
+  // Parse descriptors one-by-one from the fetched concatenated blob. Maintain
+  // pointers to the LibusbJs objects that correspond to the currently parsed
+  // interface/endpoint descriptor.
+  /** @type {!LibusbJsInterfaceDescriptor|undefined} */
+  let currentInterface = undefined;
+  /** @type {!LibusbJsEndpointDescriptor|undefined} */
+  let currentEndpoint = undefined;
+  for (let offset = 0; offset < totalLength;) {
+    const descriptorLength = descriptors.getUint8(offset);
+    if (descriptorLength < 2) {
+      // Ignore the invalid (too short) descriptor.
+      break;
```

Comment:
This should probably be `continue`.

Reply:
Guys The Amazing work  you have no idea how much you have helped

Reply:
how can I give you access my trace Logs